### ISSUE TITLE
[Bug] NemotronH --mamba-scheduler-strategy extra_buffer accuracy drop on AIME26 (Nemotron-3-Super-120B)

### ISSUE THREAD

**Issue by b8zhong** [sgl-project/sglang#31833]
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Bug Feature Tracking for NemotronH `--mamba-scheduler-strategy extra_buffer`:

| mamba-scheduler-strategy | attention-backend | AIME26 pass@1 (avg-of-16, n=30x16) |
|---|---|---|
| `no_buffer` (default) | (auto) | 84.38% ± 3.59% |
| `extra_buffer` | `trtllm_mha` | 85.21% ± 3.44% — **overlaps baseline, no regression observed (see update below)** |

### Describe the bug

Serving `nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-BF16` (NemotronH, hybrid Mamba2 + attention) with `--mamba-scheduler-strategy extra_buffer` produces a visible accuracy drop on AIME26 vs. the default `no_buffer` strategy, with everything else (model, TP/EP layout, parsers, eval config) held identical. `extra_buffer` is meant to be a pure scheduling/caching optimization (overlap schedule + mamba-state radix-cache branching) with no expected effect on generation quality.

Root cause is suspected to be in the mamba-state tracking/checkpointing math in `hybrid_linear_attn_backend.py::_init_track_ssm_indices`, used by `Mamba2AttnBackend` for this model's SSM layers:

`_init_track_ssm_indices` computes a per-request "how many packed intermediate `h` states did this request produce" count as `extend_seq_lens // mamba_cache_chunk_size` and cumulative-sums it across requests to find each request's offset into the packed `intermediate_states` tensor returned by `mamba_chunk_scan_combined`. That tensor is actually indexed by a **global physical chunk grid over the whole flattened extend batch** (`ssd_state_passing.py`'s `_state_passing_fwd_kernel` loops `for c in range(nchunks)` over the *shared* grid, writing one state per physical chunk regardless of request boundaries), not by a per-request-local contiguous count. The cumulative local-count offset is only correct if every earlier request in the same tracked prefill batch has a chunk-aligned length — not guaranteed in general.

For this model (`chunk_size=128` in `config.json`, i.e. native `mamba_chunk_size=128`) served at `--page-size 64` (the effective default when radix cache + `extra_buffer` + `trtllm_mha` are combined), half of all page-aligned tracked prefill lengths are *not* chunk-128-aligned, so the affected `_init_track_ssm_indices` "unaligned" branch fires routinely — this is not an edge case.

The net effect: a wrong (but same-shaped) SSM state gets read from `intermediate_states` and written into a request's mamba radix-cache slot whenever this happens, silently corrupting later prefix-cache-hit continuations for that node — no crash, no assertion, just wrong logits from that point on. This matches the observed AIME26 score degradation: `extra_buffer` should be numerically a no-op vs. `no_buffer`, and any drop attributable to it (beyond noise/seed variance) is evidence of this corruption path.

_(Score for the `extra_buffer` run is pending — will update this issue once the run completes.)_

### Reproduction

Server (baseline — accuracy OK, `no_buffer` is the resolved default):

```
python3 -m sglang.launch_server \
  --model-path nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-BF16 \
  --trust-remote-code \
  --tp 4 \
  --ep 4 \
  --reasoning-parser nemotron_3 \
  --tool-call-parser qwen3_coder
```

Server (suspected buggy — `extra_buffer` mamba radix-cache strategy):

```
python3 -m sglang.launch_server \
  --model-path nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-BF16 \
  --trust-remote-code \
  --tp 4 \
  --ep 4 \
  --reasoning-parser nemotron_3 \
  --tool-call-parser qwen3_coder \
  --mamba-scheduler-strategy extra_buffer \
  --attention-backend trtllm_mha
```

Eval (identical for both):

```
sgl-eval run aime26 --base-url http://localhost:30000/v1 --n-repeats 16 --max-tokens 64000 --temperature 1.0 --top-p 0.95 --num-threads 416
```

Baseline (`no_buffer`) result:

```
== aime26 ==
30 examples x 16 repeats  |  1384.2s  |  7824 tok/s  |  10.8M tokens

* pass@1[avg-of-16]  =  84.38% +/- 3.59% (SEM 0.90%)
  pass@16            =  93.33%
  majority@16        =  90.00%
  no_answer          =  11.88%  [warn: consider --max-tokens]
  stop_rate          =  88.12%
  truncated_rate     =  11.88%  [warn: hitting max_tokens]
  error_rate         =  0.00%
```

`extra_buffer` result: **TBD**

### Environment

```
Python: 3.12.3 (main, Mar 23 2026, 19:04:32) [GCC 13.3.0]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: NVIDIA B300 SXM6 AC
GPU 0,1,2,3,4,5,6,7 Compute Capability: 10.3
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 13.0, V13.0.88
CUDA Driver Version: 610.43.02
PyTorch: 2.11.0+cu130
sglang: 0.5.15.dev1093+ge856eae92
sglang-kernel: 0.4.5
flashinfer_python: 0.6.14
flashinfer_cubin: 0.6.14
flashinfer_jit_cache: 0.6.14+cu130
triton: 3.6.0
transformers: 5.12.1
```

**Comment by b8zhong** [comment]
### Update: `extra_buffer` result is in — does NOT show an accuracy regression

```
== aime26 ==
30 examples x 16 repeats  |  1368.6s  |  7854 tok/s  |  10.7M tokens

* pass@1[avg-of-16]  =  85.21% +/- 3.44% (SEM 0.86%)
  pass@16            =  96.67%
  majority@16        =  93.33%
  no_answer          =  12.71%  [warn: consider --max-tokens]
  stop_rate          =  87.50%
  truncated_rate     =  12.50%  [warn: hitting max_tokens]
  error_rate         =  0.00%
```

`extra_buffer`: 85.21% ± 3.44%, vs. baseline (`no_buffer`) 84.38% ± 3.59%. The two intervals overlap almost entirely and the point estimate is if anything slightly *higher* for `extra_buffer` — this is noise, not a regression. **I'm retracting the accuracy-drop claim**: this specific test does not demonstrate the effect described above.

Two likely reasons this test doesn't discriminate:
1. AIME26's 30 problems share essentially no prefix beyond a system prompt, so there may be little to no actual mamba radix-cache branching/reuse happening in this workload regardless of concurrency — the suspected `_init_track_ssm_indices` offset defect only manifests when a cached mamba-state node is revisited at a non-chunk-aligned boundary with other concurrently-tracked requests in the same batch, which this benchmark likely doesn't exercise.
2. n=480 with SEM ≈ 3.5% has limited power to detect a small effect either way.

The code-level observation in the `_init_track_ssm_indices` offset math (global physical chunk grid vs. per-request cumulative local count) may still be worth a second look independently, but it should be treated as an unconfirmed static-reading suspicion, not an established bug with empirical accuracy evidence — this benchmark doesn't back that claim.

**Comment by adityakamat24** [comment]
Hey @b8zhong , can I investigate it?

**Comment by b8zhong** [comment]
Sure. But I'm currently collecting more results, as the AIME result isn't conclusive right now. But I am 95% certain there's a bug still

**Comment by adityakamat24** [comment]
Thanks! I’ll also start digging into it and share whatever I find here.

**Comment by b8zhong** [comment]
Let's fully analyze the diff of https://github.com/sgl-project/sglang/pull/24955 as well, because it already introduced the earlier bug, so it might have further issues. But it might also apply to all Mamba like Q3.5

**Comment by b8zhong** [comment]
### Additional data point: GPQA diamond, extra_buffer vs no-extra-buffer

Ran a second benchmark (GPQA diamond instead of AIME26, n=198x8) to add another data point on whether `extra_buffer` causes a real accuracy regression:

```
sgl-eval run gpqa --base-url http://localhost:30000/v1 --num-examples 198 --n-repeats 8 --max-tokens 131000 --temperature 1.0 --top-p 0.95 --num-threads 416
```

**With `extra_buffer`:**
```
== gpqa ==
198 examples x 8 repeats  |  5090.0s  |  6787 tok/s  |  34.5M tokens

* pass@1[avg-of-8]  =  81.38% +/- 1.39% (SEM 0.49%)
  pass@8            =  94.44%
  majority@8        =  84.85%
  no_answer         =  1.96%
  stop_rate         =  98.11%
  truncated_rate    =  1.89%  [warn: hitting max_tokens]
  error_rate        =  0.00%
```

**Without extra_buffer (`no_buffer`, default):**
```
== gpqa ==
198 examples x 8 repeats  |  5258.5s  |  6545 tok/s  |  34.4M tokens

* pass@1[avg-of-8]  =  80.74% +/- 0.99% (SEM 0.35%)
  pass@8            =  94.44%
  majority@8        =  85.35%
  no_answer         =  1.83%
  stop_rate         =  98.17%
  truncated_rate    =  1.83%  [warn: hitting max_tokens]
  error_rate        =  0.00%
```

Like the AIME26 result above, `extra_buffer` (81.38%) is not lower than the baseline (80.74%) here either — if anything slightly higher, well within 1 SEM. pass@8 is identical between the two and majority@8 is actually a bit higher for `no_buffer`.

So across two different eval sets now (AIME26 and GPQA), I haven't been able to get an eval to actually demonstrate the accuracy drop empirically, despite the suspected offset bug in `_init_track_ssm_indices` still looking real on a code-reading basis. Consistent with what @adityakamat24 and I are digging into re: whether the affected code path is even exercised by these benchmarks (low mamba radix-cache branching / cache-hit continuation reuse in either GPQA or AIME26 workloads), vs. a benchmark power issue, vs. the bug not actually manifesting the way I suspected.

**Comment by adityakamat24** [comment]
for Mamba2 the `h` tensor from `mamba_chunk_scan_combined` isn't split per request. `nchunks` is `ceil(total_tokens / chunk_size)` over the whole flattened batch, so `h[c]` is the state at flat token `c * chunk_size` and chunks cross request boundaries. But `_init_track_ssm_indices` indexes it like every request has its own grid. The state we want for request i sits at `extend_start_loc[i] + aligned_len`, and that's only a real chunk boundary if `extend_start_loc[i] % chunk_size == 0`.

So the first request in a prefill batch is fine and the rest read the state from an earlier boundary. No crash because it's still in range.

I wrote a small script that just calls the triton ops. For ground truth I rerun each request on its own, cut to the length we claim to snapshot, and take its `varlen_state`. On an RTX 5080:

```
lens=[500,700,900] chunk=128
  req0  reads flat 384   should be 384    err 0
  req1  reads flat 1024  should be 1140   err 104%
  req2  reads flat 1920  should be 2096   err 214%

lens=[256,300,400] chunk=128
  req1  reads flat 512   should be 512    err 0
  req2  reads flat 896   should be 940    err 135%
```

Anything starting chunk aligned is exact, anything else is way off. Single request batches always pass, which is probably why it never showed up.

On #24955, I don't think that's the one. `_init_track_ssm_indices` came from #14792 (Qwen3-Next) and it was correct there. #15829 is what added the `isinstance(self, Mamba2AttnBackend)` branch when `extra_buffer` was extended to Mamba2 , it changed ceil to floor but kept the per request cumsum, which doesn't hold for Mamba2's layout.

On Q3.5, I don't think it's affected. Q3.5 and Qwen3-Next go through `GDNAttnBackend`, and FLA does chunkoffsets is `cumsum(ceil(len/64))`), so the offset math is fine there. It's only the Mamba2 side that'swrong.

One more thing I noticed, `_init_track_conv_indices` uses the exact flat position (`query_start_loc[i] + aligned_len`), so conv and ssm in the same cache entry end up snapshotted at different token counts.

**Comment by adityakamat24** [comment]
@b8zhong I fixed it. Instead of indexing into `h`, I now pass the flat positions we actually want
and let `chunk_state_varlen` rebuild the exact state there. The FLA path is untouched.

At the kernel level I checked it against unmodified main, using each request rerun on its own,
truncated to the snapshot length, as the reference. Before the fix the misaligned reads were 100%+
off, after it they're exact. The existing suite went from 364 to 370 passing, no regressions.

Then end to end on a single H100 with Nemotron-Nano-9B-v2, same NemotronH arch and same
`chunk_size: 128` as the 120B. GSM8K Platinum, 400 questions x 8 repeats, `--parallel 128`,
temperature 1.0, top-p 0.95:

```
no fix  : avg-of-8 = 84.86% +/- 1.42%  (SEM 0.73%)  min 82.0  max 88.2
with fix: avg-of-8 = 84.92% +/- 0.82%  (SEM 0.42%)  min 83.3  max 86.5
delta   = +0.06%  (t = 0.07)
```

So accuracy doesn't move, same as what you saw on AIME26.

The generated text does change though. I took 200 prompts at concurrency 64 and measured the noise
floor first by rerunning the same build:

```
fixed vs fixed     (same launch and across launches) : 20-30%
pre-fix vs pre-fix (across launches)                 : 43.5%
pre-fix vs fixed                                     : 94.5-96%
```

Both builds are far more consistent with themselves than they are with each other, so that 95% is
systematic rather than run-to-run noise. And since chunk-aligned requests are bit-identical between
old and new code, the misaligned ones are the only thing that can be driving it. The other thing
that shows up is that the fix makes output about twice as reproducible, which fits, since which
requests end up misaligned depends on batch composition. The narrower accuracy spread above is the
same effect.

So the state really is wrong and it does change what the model writes, it just doesn't change
whether the answer is right. NemotronH is hybrid, so the attention layers still hold a correct KV
cache over the tokens the mamba state dropped and the model gets to the same answer another way.
That's also why AIME26 couldn't pick this up, task accuracy is too coarse for it, the text differs
but the final answers mostly don't.

**Comment by adityakamat24** [comment]
@b8zhong you were right, there is a bug. I raised a PR #33450 with a fix. Please do take a look whenever you get time, thanks!

`_init_track_ssm_indices` indexes the Mamba2 per-chunk `h` as a per-request grid, but it's one global grid over the flattened batch, so any request that doesn't start chunk-aligned caches the wrong SSM state. Against unmodified main it's bit-exact for aligned requests and 100%+ off for the rest. Accuracy is unchanged (matches AIME26), but under chunked prefill the reasoning changes on ~90% of prompts against a ~15% floor.

**Comment by b8zhong** [comment]
Seems the fix is from BCG
